### DISCUSSION TITLE
CTkTable Widget

### DISCUSSION THREAD

**Discussion by Akascape** [TomSchimansky/CustomTkinter#1483] [Showcase]
Here is a quick and simple table widget having all the basic features. 

![Screenshot](https://user-images.githubusercontent.com/89206401/233420929-bf210cb3-5b5f-49b2-ba7a-f01d187e72cf.jpg)

## Features:
- Add columns/rows
- Delete columns/rows
- Edit rows/columns at once
- Insert values to specific block
- delete values from specific block
- update all values at once
- customize each block and fg-colors
- lightweight widget class
- can be used with scrollable frame

Download from here: https://github.com/Akascape/CTkTable
This is a updated version of ctktable I made, use it if you want to deal with **basic table operations.**

**Comment by avalon60** [comment]
Sweet. Will you be making this available via pip?

**Reply by Akascape** [reply]
@avalon60 Not for now. It lacks many functionalities like adding or deleting row/columns, updating colors etc.

**Reply by avalon60** [reply]
Would be useful to display icons too. Not sure how easy that would be to implement though.

**Reply by Akascape** [reply]
@avalon60 
Yes you can do that as every block is a button, just pass any button argument in `insert(...)`.

**Reply by ghost** [reply]
When using buttons to display the images, you want to disable their hover effect either through the `disabled` status or through the theming.

**Comment by RickWalkerOne** [comment]
Hello! Man, you are a top programmer! I love you widgets!

Its beautfull! But where is the CTkTable not simple that you created?

**Reply by PaulHalbig** [reply]
@Akascape just wanted to bump this in case you forgot it. A fully featured table widget would be awesome.

**Reply by Akascape** [reply]
I am working on it. Will be available soon.

**Reply by RickWalkerOne** [reply]
@Akascape
If help you, I made a datagridview for custom tkinter, maybe help you

**Reply by Akascape** [reply]
@RickWalkerOne @PaulHalbig 
I have just updated it, now it contains almost all the required features.

**Comment by VasigaranAndAngel** [comment]
You're fantastic with these tools! I'm a fan of your projects. what do you think about animating the widgets? I wonder if it's possible to animate without any struggles.

**Reply by Akascape** [reply]
@VasigaranAndAngel 
> You're fantastic with these tools! I'm a fan of your projects. 

Thanks for the feedback :)

> what do you think about animating the widgets? I wonder if it's possible to animate without any struggles

Animating widget is a bit tricky with canvas type widgets, you have to place them properly and update simultaneously. You can also configure the padding if you are using grid/pack method.

**Reply by avalon60** [reply]
@Akascape, I see this looks more complete now, any plans to publish via PyPi?

**Reply by VasigaranAndAngel** [reply]
> Animating widget is a bit tricky with canvas type widgets, you have to place them properly and update simultaneously. You can also configure the padding if you are using grid/pack method.

Will it be difficult? Can't we try?

**Comment by learningtocodern** [comment]
Hey man, your widgets are awesome.
I want to ask if there is any method to export ctktable directly to pdf or indirectly to pdf using python modules??

**Reply by Akascape** [reply]
@learningtocodern Like a screenshot or in a classic table?

**Reply by learningtocodern** [reply]
classic table?

**Reply by Akascape** [reply]
@learningtocodern yeah, I mean to say, do you want the data from the ctk table moved to a simple spreadsheet and convert to pdf?

**Reply by learningtocodern** [reply]
Yes that's the method I want

**Reply by learningtocodern** [reply]
can you suggest any method for the same?

**Reply by ghost** [reply]
@learningtocodern - So the data will be moved to a spreadsheet, and then into a table in a PDF. The PDF part should be possible with the fpdf module. For the spreadsheet, maybe use openpyxl. This works with Microsoft Excel Spreadsheets.

**Comment by jklouz** [comment]
how to set text alignment in cell?

**Reply by Akascape** [reply]
@jklouz pass `anchor` parameter.

**Reply by jklouz** [reply]
thats only works for direct string value, but not from variable like displaying contents of text from separate file, how to solve this?

table.add_row([dictionary['key']],anchor='w') # imported text not anchored
table.add_row(['direct strings'],anchor='w')   # direct text is anchored

**Reply by Akascape** [reply]
@jklouz Please refer the documentation. We have functions for all operations. You have to pass only list of values in ctktable.

**Comment by Dumbliidore** [comment]
The image cannot be displayed.
This is my code
```python
import customtkinter
from CTkTable import *
from PIL import Image

root = customtkinter.CTk()

button = customtkinter.CTkButton(root, text="")
button.configure(image=customtkinter.CTkImage(Image.open(r"D:\\Study\\translator\\checkmark.png")))
button.pack(expand=True, fill="both")

value = [[1,2,3,4,5],
         [1,2,3,4,5],

         [1,2,3,4,5],
         [1,2,3,4,5],
         [1,2,3,4,5]]

table = CTkTable(master=root, row=5, column=5, values=value)
table.insert(1, 2, value="", image=customtkinter.CTkImage(Image.open(r"D:\Study\translator\checkmark.png")))
table.pack(expand=True, fill="both", padx=20, pady=20)

root.mainloop()
```
![image](https://github.com/TomSchimansky/CustomTkinter/assets/35067857/372a550a-d9ec-44ca-a653-5221b1ddc7af)

**Reply by Akascape** [reply]
@D-Ayden Update to the latest version. Its fixed now.

**Reply by Dumbliidore** [reply]
It's good for me, thanks.
